### ISSUE TITLE
Easy to get needle-like Gaussians

### ISSUE THREAD

**Issue by jiahaolu97** [pointrix-project/msplat#6]
I transfered the whole optimization and dataset/camera reading pipeline into this repo, and I found that with the same hyper-parameter setting as original 3DGS repo default, it will easily produce needle-like gaussians:

<img width="551" alt="289bc80c7071a24624376da3d3a4a79" src="https://github.com/pointrix-project/msplat/assets/38347064/0c1451c0-3373-4dc9-844c-5e63eea776cf">

Could you check if there is any errors in kernel computation (likely the backward computation of scale)?  I can share my transferral of other components (optimization pipeline, 3D dataset readers) here if needed (likely some time after NeurIPS ddl).

**Comment by yGaoJiany** [comment]
🤔 It's a little strange. I've also tried training NeRF _Lego_ with msplat, but haven't found this problem.

**Comment by jiahaolu97** [comment]
Emmm that's really weird. I carefully transfer the code from original 3DGS repo so the hyperparameter settings are the same ...

May I ask if there is any change in hyper-parameters in your case?

Since from my observation, using default hyper-parameters here will result fewer Gaussians (original: ~600 K gaussians, here: ~5K Gaussians ) - seems only a few of Gaussians reached the densification threshold and got densified.

**Comment by yGaoJiany** [comment]
Well, I will check this 🏗️. Please keep this issue open!

**Comment by jiahaolu97** [comment]
Thank you! Glad to discuss more on this and provide potential assistance.

**Comment by yGaoJiany** [comment]
Well, I did a quick replacement based on the original repo of gaussian splatting. And it seems that msplat works just fine.  Here is what I get at 7000 iteration.
![image](https://github.com/pointrix-project/msplat/assets/36816705/c1143c9f-0a74-4102-b996-7178ce6a5041)

**Comment by yGaoJiany** [comment]
I replaced the function named as _render_ in _gaussian_renderer/__init__.py_ with this one:
```
import msplat as ms

def render(viewpoint_camera, pc : GaussianModel, pipe, bg_color : torch.Tensor, scaling_modifier = 1.0, override_color = None):
    """
    Render the scene with msplat.
    
    Background tensor (bg_color) must be on GPU!
    """
    
    # print("Hello, masplat.")
    
    # tranform 3dgs to msplat
    position = pc.get_xyz
    fovx = viewpoint_camera.FoVx
    fovy = viewpoint_camera.FoVy
    width = int(viewpoint_camera.image_width)
    height = int(viewpoint_camera.image_height)
    
    fx = fov2focal(fovx, width)
    fy = fov2focal(fovy, height)
    cx = float(width) / 2
    cy = float(height) / 2
    
    intrinsic_params = torch.tensor([fx, fy, cx, cy]).cuda().float()
    extrinsic_matrix = viewpoint_camera.world_view_transform.transpose(0, 1)
    extrinsic_matrix = extrinsic_matrix[:3, :]
    camera_center = viewpoint_camera.camera_center
    
    opacity = pc.get_opacity
    shs = pc.get_features.permute(0, 2, 1)
    scaling = pc.get_scaling
    rotation = pc.get_rotation

    # project points and perform culling
    (uv, depth) = ms.project_point(
        position,
        intrinsic_params,
        extrinsic_matrix,
        width, height)

    visible = depth != 0

    # compute sh if not None
    direction = (position -
                camera_center.repeat(position.shape[0], 1))
    direction = direction / direction.norm(dim=1, keepdim=True)
    
    sh2rgb = ms.compute_sh(shs, direction, visible)
    rgb = torch.clamp_min(sh2rgb + 0.5, 0.0)
    
    # compute cov3d
    cov3d = ms.compute_cov3d(scaling, rotation, visible)

    # ewa project
    (conic, radius, tiles_touched) = ms.ewa_project(
        position,
        cov3d,
        intrinsic_params,
        extrinsic_matrix,
        uv,
        width,
        height,
        visible
    )

    # sort
    (gaussian_ids_sorted, tile_range) = ms.sort_gaussian(
        uv, depth, width, height, radius, tiles_touched
    )
    
    # render
    ndc = torch.zeros_like(uv, requires_grad=True)
    try:
        ndc.retain_grad()
    except:
        raise ValueError("ndc does not have grad")

    # alpha blending
    render = ms.alpha_blending(
        uv, conic, opacity, rgb,
        gaussian_ids_sorted, tile_range, bg_color[0].item(), width, height, ndc
    )
    
    return {"render": render,
            "viewspace_points": ndc,
            "visibility_filter" : radius > 0,
            "radii": radius}
```

**Comment by yGaoJiany** [comment]
Here's what I got with msplat
```
[ITER 30000] Evaluating test: L1 0.004826885849470273 PSNR 36.27218983650208 [17/05 16:02:35]
[ITER 30000] Evaluating train: L1 0.0029831654857844117 PSNR 40.72242660522461 [17/05 16:02:35]
```
The number of points is 345,685.
![image](https://github.com/pointrix-project/msplat/assets/36816705/209e6f68-35f6-4fda-9058-840b920bc04d)

**Comment by yGaoJiany** [comment]
Here is the report for original 3DGS.
```
[ITER 30000] Evaluating test: L1 0.004817731926450506 PSNR 36.26937942504883 [17/05 16:18:35]
[ITER 30000] Evaluating train: L1 0.0029933738056570295 PSNR 40.61189193725586 [17/05 16:18:35]
```
The number of point is 343,412.

It seems that msplat is a little bit slower. For _Lego_, msplat costs 10min12s for 30000 iteration while the original kernel costs 7min13s.

**Comment by jiahaolu97** [comment]
Thank you @yGaoJiany  for providing your results! 

Let me check on my codes. I will get back to this issue after some time to report where my bugs are and hope it can help more users.

**Comment by Yuhuoo** [comment]
I think it may be a data problem. When I used 30 images to train the Horse model using mslpat, it quickly overfitted after about 3000 iterations. The loss started increasing, and after training, the rendered results were very bad. However, I didn't encounter this problem in the native project of Graphdeco-Inria. Could there be an unstable gradient calculation in this msplat project?

Here is the log of training.
```
(gaussian_splatting) aogao@test-X640-G40:~/code/dust_gs/gaussian-splatting$  cd /home/aogao/code/dust_gs ; /usr/bin/env /usr/bin/python3 /data1/GaoAo/data/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher 60239 -- /home/aogao/code/dust_gs/gaussian-splatting/train.py -s /home/aogao/code/dust_gs/gaussian-splatting/data/Horse_30 -m /home/aogao/code/dust_gs/gaussian-splatting/output/Horse_30_instantsplat_wo10000_wTl --iterations 7000 
Optimizing /home/aogao/code/dust_gs/gaussian-splatting/output/Horse_30_instantsplat_wo10000_wTl
Output folder: /home/aogao/code/dust_gs/gaussian-splatting/output/Horse_30_instantsplat_wo10000_wTl [17/05 17:48:27]
Reading camera 30/30 [17/05 17:48:27]
Generating ellipse path from 30 camera infos ... [17/05 17:48:27]
theta[0] 0.0 [17/05 17:48:27]
Loading Training Cameras [17/05 17:49:26]
Loading Test Cameras [17/05 17:49:33]
Loading Render Cameras [17/05 17:49:33]
Number of points at initialisation :  1289775 [17/05 17:49:34]
Training progress:  14%|███████▌                                             | 1000/7000 [01:03<06:05, 16.42it/s, Loss=0.0647089]
[ITER 1000] Evaluating train: L1 0.03905064985156059 PSNR 21.436789321899415 [17/05 17:50:40]

[ITER 1000] Saving Gaussians [17/05 17:50:40]
Training progress:  53%|████████████████████████████                         | 3700/7000 [03:52<02:06, 26.05it/s, Loss=0.5564653]
```

When I use all the Horse data (120 images) with msplat for training, the training is stable, and the results are resonable.
```
(gaussian_splatting) aogao@test-X640-G40:~/code/dust_gs/gaussian-splatting$ CUDA_VISIBLE_DEVICES=1 python train.py -s /home/aogao/code/dust_gs/gaussian-splatting/data/tanks_temples/Horse -m /home/aogao/code/dust_gs/gaussian-splatting/output/Horse --eval
Optimizing /home/aogao/code/dust_gs/gaussian-splatting/output/Horse
Output folder: /home/aogao/code/dust_gs/gaussian-splatting/output/Horse [17/05 18:21:58]
Reading camera 120/120 [17/05 18:21:59]
Generating ellipse path from 120 camera infos ... [17/05 18:21:59]
theta[0] 0.0 [17/05 18:21:59]
Loading Training Cameras [17/05 18:21:59]
Loading Test Cameras [17/05 18:22:07]
Loading Render Cameras [17/05 18:22:07]
Number of points at initialisation :  12141 [17/05 18:22:09]
Training progress:   3%|█▋                                                  | 1000/30000 [00:39<16:56, 28.54it/s, Loss=0.1091999]
[ITER 1000] Evaluating test: L1 0.05780421371261279 PSNR 19.022645696004233 [17/05 18:22:51]

[ITER 1000] Evaluating train: L1 0.05654826983809472 PSNR 19.250444793701174 [17/05 18:22:54]

[ITER 1000] Saving Gaussians [17/05 18:22:54]
Training progress:  23%|████████████▏                                       | 7000/30000 [04:02<13:20, 28.74it/s, Loss=0.0770475]
[ITER 7000] Evaluating test: L1 0.04094977639615536 PSNR 20.788012568155924 [17/05 18:26:13]

[ITER 7000] Evaluating train: L1 0.03682942055165768 PSNR 21.61775703430176 [17/05 18:26:14]

[ITER 7000] Saving Gaussians [17/05 18:26:14]
Training progress: 100%|███████████████████████████████████████████████████| 30000/30000 [18:32<00:00, 26.96it/s, Loss=0.0524706]

[ITER 30000] Evaluating test: L1 0.03810176663100719 PSNR 21.380385208129884 [17/05 18:40:43]

[ITER 30000] Evaluating train: L1 0.02782416120171547 PSNR 23.820465087890625 [17/05 18:40:44]

[ITER 30000] Saving Gaussians [17/05 18:40:45]

Training complete. [17/05 18:40:50]
```

**Comment by jiahaolu97** [comment]
@Yuhuoo  I don't think we have the same problem, since I tested for the same data between `msplat` and the Inria repo ... Using 100 or 300 camera views both produce needle like Gaussians in my implementation with `msplat`, however in the Inria repo I didn't meet the same problem (either 100 images or 300 images).

Given the information from @yGaoJiany, I assume there's maybe some bugs in my code, not with their cuda kernel. I need to have some time to check on this though.

**Comment by yGaoJiany** [comment]
@Yuhuoo Maybe you can start a new issue about overfitting in a setting of fewer images and we can discuss this topic there.
